pkg/utils/filter_map.go

```diff
@@ -0,0 +1,45 @@
+/*
+Copyright © 2020 Doppler <support@doppler.com>
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package utils
+
+import (
+	"strings"
+)
+
+// FilterMap will return either:
+//
+// map[string]T, nil where map[string]T contains all elements where the key matched a key inside keysToSelect
+//
+// nil, []string where []string contains the keys that did not match to a key inside origMap
```

Comment:
The dual return types are a bit confusing to me. Would this be better accomplished with two discrete functions? Though given that this is a filter, I'm curious why it wouldn't just return the newly filtered map? Lmk if you want to chat about this further offline.

You'll also want to prefix this commit with `chore:`.

Reply:
The reason I went for a dual return is to provide extra information on the filtering result. 
The first return value contains the new map that has contains items the items present in the array of keys provided as the second parameter of the function
The second return potentially contains an array of keys that weren't present in the original map but asked to be filtered out by the user. (non existent keys use case)

This allows us to then check the second return value and craft the error message for the user:
https://github.com/DopplerHQ/cli/blob/c5d8c531f27d3bbd6855e71bbcb18afc1ba7ce21/pkg/controllers/secrets.go#L212-L214

How would you envision crafting the error message without the information of keys that were asked to be included but don't exist in the secret map?

Reply:
I think that should be the responsibility of the controller function. It would loop through the newly filtered map and find the keys that don't exist. I don't think that should be the responsibility of our generic FilterMap util though. I'd expect that to have an obvious, somewhat standard interface.